tests/unit/base/trained_model_serialization_unittest.cc

```diff
@@ -0,0 +1,197 @@
+#include "environments/LinearTestEnvironment.h"
+#include "environments/MultiLabelTestEnvironment.h"
+#include "environments/RegressionTestEnvironment.h"
+#include "utils/Utils.h"
+#include <gtest/gtest.h>
+#include <shogun/base/some.h>
+#include <shogun/features/DenseFeatures.h>
+#include <shogun/io/CSVFile.h>
+#include <shogun/io/SGIO.h>
+#include <shogun/io/SerializableAsciiFile.h>
+#include <shogun/io/SerializableHdf5File.h>
+#include <shogun/kernel/GaussianKernel.h>
+#include <shogun/labels/BinaryLabels.h>
+#include <shogun/labels/Labels.h>
+#include <shogun/labels/RegressionLabels.h>
+#include <shogun/machine/Machine.h>
+
+using namespace shogun;
+
+extern LinearTestEnvironment* linear_test_env;
+extern MultiLabelTestEnvironment* multilabel_test_env;
+extern RegressionTestEnvironment* regression_test_env;
+
+template <class T>
+class TrainedModelSerializationFixture : public ::testing::Test
+{
+protected:
+	void SetUp()
+	{
+		machine = new T();
+		SG_REF(machine)
+
+		deserialized_machine = new T();
+		SG_REF(deserialized_machine)
+
+		this->load_data(this->machine->get_machine_problem_type());
+	}
+
+	void TearDown()
+	{
+		SG_UNREF(train_feats)
+		SG_UNREF(test_feats)
+		SG_UNREF(train_labels)
+		SG_UNREF(machine)
+		SG_UNREF(deserialized_machine)
+	}
+
+	void load_data(EProblemType pt)
+	{
+		switch (pt)
+		{
+		case PT_BINARY:
+		case PT_CLASS:
+		{
+			std::shared_ptr<GaussianCheckerboard> mock_data =
+			    linear_test_env->getBinaryLabelData();
+			train_feats = mock_data->get_features_train();
+			test_feats = mock_data->get_features_test();
+			train_labels = mock_data->get_labels_train();
+			break;
+		}
+
+		case PT_MULTICLASS:
+		{
+			std::shared_ptr<GaussianCheckerboard> mock_data =
+			    multilabel_test_env->getMulticlassFixture();
+			train_feats = mock_data->get_features_train();
+			test_feats = mock_data->get_features_test();
+			train_labels = mock_data->get_labels_train();
+			break;
+		}
+
+		case PT_REGRESSION:
+			train_feats = regression_test_env->get_features_train();
+			test_feats = regression_test_env->get_features_test();
+			train_labels = regression_test_env->get_labels_train();
+			break;
+
+		default:
+			SG_SERROR("Unsupported problem type: %d\n", pt);
+			FAIL();
+		}
+
+		SG_REF(train_feats)
+		SG_REF(test_feats)
+		SG_REF(train_labels)
+	}
+
+	bool serialize_machine(
```

Comment:
shall we put this into a helper, this kind of stuff is used in multiple places in the tests